src/pke/lib/scheme/ckksrns/ckksrns-leveledshe.cpp

```diff
@@ -360,15 +360,52 @@ std::vector<DCRTPoly::Integer> LeveledSHECKKSRNS::GetElementForEvalAddOrSub(
   const auto cryptoParams =
       std::static_pointer_cast<CryptoParametersCKKSRNS>(
           ciphertext->GetCryptoParameters());
-  double scFactor =
-      cryptoParams->GetScalingFactorReal(ciphertext->GetLevel());
 
-  DCRTPoly::Integer intScFactor = static_cast<uint64_t>(scFactor + 0.5);
+  double scFactor;
+  if (cryptoParams->GetRescalingTechnique() == FLEXIBLEAUTOEXT && ciphertext->GetLevel() == 0) {
+    scFactor = cryptoParams->GetScalingFactorRealBig(ciphertext->GetLevel());
+  } else {
+    scFactor = cryptoParams->GetScalingFactorReal(ciphertext->GetLevel());
+  }
+
+  // Compute approxFactor, a value to scale down by, in case the value exceeds a 64-bit integer.
+  int32_t MAX_BITS_IN_WORD = 62;
+
+  int32_t logSF = static_cast<int32_t>(ceil(log2(fabs(scFactor))));
+  int32_t logValid = (logSF <= MAX_BITS_IN_WORD) ? logSF : MAX_BITS_IN_WORD;
+  int32_t logApprox = logSF - logValid;
+  double approxFactor = pow(2, logApprox);
+
   DCRTPoly::Integer scConstant =
-      static_cast<uint64_t>(constant * scFactor + 0.5);
+      static_cast<uint64_t>(constant * scFactor / approxFactor + 0.5);
+  std::vector<DCRTPoly::Integer> crtConstant(sizeQl, scConstant);
 
+  // Scale back up by approxFactor within the CRT multiplications.
+  int32_t MAX_LOG_STEP = 60;
```

Comment:
is it possible to define MAX_LOG_STEP in one place?

Reply:
Defined it in constants.h